### DISCUSSION TITLE
How can I get raw text output from my WebSocket connection via client requests?

### DISCUSSION THREAD

**Discussion by mfreidel** [evennia/evennia#2570] [Evennia Questions] [answered]
I've been poking around at the WebSockets interface (trying to make my own external client), and I don't want any HTML in my output. In websockets.py (https://github.com/evennia/evennia/blob/master/evennia/server/portal/webclient.py), there seems to be an option to send raw text...

```
def send_text(self, *args, **kwargs):
...
    options = kwargs.pop("options", {})
    raw = options.get("raw", flags.get("RAW", False))
    client_raw = options.get("client_raw", False)
...
    if raw:
        if client_raw:
            args[0] = text
        else:
            args[0] = html.escape(text)  # escape html!
```
... However, I can't seem to get the server to do this via client requests. I have tried to send webclient_options OOB commands like webclient.js does, but they don't do anything and I don't get response packets like it does in any case.

I've tried sending the following JSON payloads after logging in:
```
["webclient_options", [], {}]
["webclient_options", [], {"options": {"raw": true}}]
["webclient_options", [], {"gagprompt": false, "helppopup": false, "notification_popup": false, "notification_sound": false, "mm_image": false, "mm_audio": false, "mm_video": false, "options": {"raw": true}}]
```

**Things of note about my client:**
- Able to connect/disconnect, login, send/receive messages via the standard WebSocket port.
- No evennia.js API
- No csessid (yet?)
- No sub-protocol declared on websocket connect. It's a pure WebSocket connection.

I'm kind of thinking it could be the sub-protocol negotiation, but I couldn't get that working either. Maybe I'm using the wrong OOB command? Not sure what I'm doing wrong here.

**Comment by Griatch** [comment, marked as answer]
The `webclient_options` is just a way for the webclient to save its own settings to the server (to the Account actually) so it can be retrieved later on a reconnect. It does not change how the server communicates to the client or the protocol flags of the protocol itself. 

To accomplish something like that, one would likely need to modify the `webclient_options` [inputfunc](https://www.evennia.com/docs/latest/Inputfuncs.html) to set a new flag in the `protocol_flags` of the webclient protocol, so as to always strip any html from the output and send raw data. 

That said ... for evennia 1.0 there is an [ongoing project](https://github.com/evennia/evennia/pull/2419) for moving the rendering of HTML from the server to the webclient. This would basically have the webclient protocol always send 'raw' data for the client to parse as it wants instead of pre-parsing to html as we do today. So once we get that project finished it should solve things for you automatically. There is no timeline on that at this time though.

**Reply by mfreidel** [reply]
I see. This is all very helpful! Thanks for pointing me in the right direction with `protocol_flags`. I think I'll keep an eye on that PR and work on my server-side connection for now then.
